internal/lib/conmonmon.go

```diff
@@ -0,0 +1,127 @@
+package lib
+
+import (
+	"sync"
+	"time"
+
+	"github.com/cri-o/cri-o/internal/oci"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+var sleepTime = 5 * time.Minute
+
+// conmonmon is a struct responsible for monitoring conmons
+// it contains a map of containers -> conmonPID, and sleeps on
+// a loop, waiting for a conmon to die. if it has, it kills the associated
+// container.
+type conmonmon struct {
+	conmons   map[*oci.Container]int
+	mu        sync.RWMutex
+	closeChan chan bool
+	runtime   *oci.Runtime
+}
+
+// newConmonmon creates a new conmonmon instance
+// given a runtime. It also starts the monitoring routine
+func newConmonmon(r *oci.Runtime) *conmonmon {
+	cmm := conmonmon{
+		conmons: make(map[*oci.Container]int),
+		runtime: r,
+	}
+	go cmm.monitorConmons()
+	return &cmm
+}
+
+// monitorConmons sits on a loop and sleeps.
+// after waking, it signals to the conmons, checking
+// if they're still alive.
+func (c *conmonmon) monitorConmons() {
+	for {
+		select {
+		case <-time.After(sleepTime):
+			c.signalConmons()
+		case <-c.closeChan:
+			return
+		}
+	}
+}
+
+// signalConmons loops through all available conmons and
+// sends a kill 0 to them, checking if they're still alive
+// if they're not, the container is killed and we spoof
+// an OOM event for the container
+func (c *conmonmon) signalConmons() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	for ctr, conmonPID := range c.conmons {
+		status := ctr.State().Status
+		if status == oci.ContainerStateRunning {
+			if err := unix.Kill(conmonPID, 0); err != nil {
+				logrus.Debugf("killing conmon pid %d returned %v. Killing container %s", conmonPID, err, ctr.ID())
+				if err := c.runtime.SignalContainer(ctr, unix.SIGKILL); err != nil {
+					logrus.Debugf(err.Error())
+				}
+				delete(c.conmons, ctr)
+				oci.SpoofOOM(ctr)
```

Comment:
We should persist the state to disk.

Reply:
Minimally, yes. I need to do more thorough test run

Reply:
You mean writing.the oom file? SGTM

Reply:
ContainerStateToDisk so we persist the state on disk. Also, we should write the oom file as well to make the state exactly as if it were an oom kill.

Reply:
fixed